Downstream data processing edits

jds485 opened the pull request:
This PR addresses 2 small issues related to downstream data processing.

1. adds a filter to `p2_all_attr_SC_obs` to remove erroneous observation documented in #190. The filter should work if we update our pipeline with the WSC-corrected data. I think we could do that update now with minimal effect on processing time, but we may discover new data issues that I'd prefer to avoid at this time. 

Closes #190 

2. adds a `data_type` column to each of the SC data sources that indicates if it is from a daily or continuous sampler. After aggregating to PRMS segments (taking the max of 'd' and 'u' to return 'u' if any days are from continuous), there are only 4209 site-days with continuous labels out of ~190k site-days. That's about equal to the number of site-days in `p2_inst_data_daily`, but I'm still wondering if some of the NWIS daily data should also receive a continuous label (i.e., because the hourly data were processed to daily within NWIS). I'm not sure if we'd have a simple way to check that.
```
length(which(p2_SC_observations$data_type == 'u'))
[1] 4209
```
Closes #191


I updated all targets before `p4` targets.

lekoenig commented on 2_process.R:
```diff
@@ -545,6 +557,9 @@ p2_targets_list <- list(
+      filter(Date >= earliest_date) %>%
+      #remove bad sample (explained here https://github.com/USGS-R/drb-inland-salinity-ml/issues/190)
+      #This filter works to remove this point if mean_value is NA or > 25000 for this Date and PRMS_segid.
+      filter(!(Date == '1999-03-30' & PRMS_segid == '111_1' & mean_value > 25000))
```
If we're already editing `p2_wqp_SC_filtered` in this PR, should we move this filtering step there? That way, we could be very explicit about which record we want to omit, rather than looking for a threshold value in `p2_all_attr_SC_obs`.

```r
... %>%
filter(!(MonitoringLocationIdentifier == "USGS-01421618" & ActivityStartDate == "1999-03-30"))
```

Either work and so I'm happy to leave this as is though.

jds485 replied:
good idea to move to `p2_wqp_SC_filtered`

lekoenig commented on 2_process/src/aggregate_observations.R:
```diff
@@ -35,11 +35,13 @@ combine_wq_data <- function(wqp_data, nwis_data){
               Value_Min = min(resultVal2, na.rm = TRUE),
               Value_Min_cd = NA_character_,
+              #using max to select 'u', which is greater than 'd'.
+              data_type = max(data_type),
```
This is a clever use of `max()`. Just to check my own understanding - are there actually any instances of 'u' in the WQP data? Based on the definition in `p2_wqp_SC_filtered` I assumed all of these would be 'd'.

jds485 replied:
They are all 'u' for WQP, so I can remove the max here. I copied this code from the combined daily + instantaneous function (also edited in this PR)

lekoenig reviewed (commented):
Thanks, @jds485 - this looks good! I just have one question that should be addressed before we merge, namely, why is 'd' assigned to both NWIS daily values and discrete grab samples? 

Could you could remind me of the motivation/use case for labeling the data types? USGS does not make public any instantaneous records prior to 2007, but my understanding is that all daily data ("dv" service) ultimately come from time series data (see the FAQ [here](https://help.waterdata.usgs.gov/faq/surface-water/surface-water-daily-data)). Discrete water quality samples usually come from a separate service, "qw." If we're specifically interested in the relative value of the "uv" data it makes sense to only have the two data types, but we should add a comment somewhere that describe what these refer to (maybe you already have that somewhere and I've just missed it). If we're interested in the value of temporal data versus spatially-distributed grab samples (which do expand the domain space of the model but usually have sparse temporal coverage), then we might consider whether the NWIS daily values should be lumped in with the "u" data type. What do you think?

jds485 commented on 2_process.R:
```diff
@@ -390,7 +400,9 @@ p2_targets_list <- list(
-    subset_wqp_nontidal(p2_wqp_SC_data,p2_sites_w_segs,mainstem_reaches_tidal)
+    subset_wqp_nontidal(p2_wqp_SC_data,p2_sites_w_segs,mainstem_reaches_tidal) %>%
+      #add data source type as daily estimate
+      mutate(data_type = 'd')
```
Okay, that's helpful. I did not know that the daily data were from ~continuous samplers. I can change the daily data to also be 'u'

lekoenig replied:
Wouldn't these be 'd' though given the code on line 405 of `2_process.R`?

jds485 replied:
yes, the WQP data are all 'd'

lekoenig replied:
Oops! I added that question in the wrong spot - I meant to reply to your comment [below](https://github.com/USGS-R/drb-inland-salinity-ml/pull/194#discussion_r1012196863).

jds485 replied:
Oh! Yes, just corrected that typo. Thanks!

jds485 commented:
> Could you could remind me of the motivation/use case for labeling the data types?

The labels are for use in spatial-only holdouts. We can try to get a representative sample of reaches with continuous samplers and w/o continuous samplers for each split (train vs. test, and within training, the analysis vs. assessment sets in CV).

> we should add a comment somewhere that describe what these refer to (maybe you already have that somewhere and I've just missed it)

I have comments in the lines above where the labels are assigned.

jds485 commented on 1_fetch.R:
```diff
@@ -94,7 +94,9 @@ p1_targets_list <- list(
     get_daily_nwis_data(p1_nwis_sites_daily, parameter, stat_cd_select,
-                        start_date = earliest_date, end_date = latest_date),
+                        start_date = earliest_date, end_date = latest_date) %>%
+      #add column to identify the data source as daily
+      mutate(data_type = 'd'),
```
I view the `p2_daily_combined` as a workaround to our current use of never rebuilding this `p1` target. But it will work just the same to assign the data_type in `p2` and not here. I can delete this

jds485 commented:
Thanks @lekoenig! I think I addressed all of your comments. The data labeled as continuous are now ~85% of the data, and those are located on ~25% of the reaches with observations. There are only 6/82 reaches with continuous data before the 2020 water year, so our prior temporal split did not have many of these reaches to train on. That definitely motivates doing spatial holdouts with these continuous-data reaches

lekoenig commented:
> I think I addressed all of your comments. The data labeled as continuous are now ~85% of the data, and those are located on ~25% of the reaches with observations. There are only 6/82 reaches with continuous data before the 2020 water year, so our prior temporal split did not have many of these reaches to train on. That definitely motivates doing spatial holdouts with these continuous-data reaches

Fascinating! Thanks for looking into this and for adding these changes.
